File: tests/vulnerabilities/go/realworld_open_redirect.go
package main

// Real-world open redirect test cases for Kunlun-M scanner
// Simulates a Go web application with login redirect and OAuth callback features
// Expected detections: 3 TRUE POSITIVE (open redirect), 1 SAFE

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

// AuthServer simulates an authentication service with redirect flows
type AuthServer struct{}

// VULN-1: Open redirect via http.Redirect with user-controlled "next" param
// User can redirect to any external URL after login
func (as *AuthServer) LoginRedirectHandler(w http.ResponseWriter, r *http.Request) {
	username := r.FormValue("username")
	password := r.FormValue("password")
	nextURL := r.URL.Query().Get("next")
	// VULN: user-controlled nextURL used directly in redirect
	// Attacker can set next=http://evil.com to phish credentials
	if username == "admin" && password == "admin123" {
		if nextURL != "" {
			http.Redirect(w, r, nextURL, http.StatusFound)
		} else {
			http.Redirect(w, r, "/dashboard", http.StatusFound)
		}
	} else {
		http.Error(w, "invalid credentials", 401)
	}
}

// VULN-2: Open redirect via http.Redirect with user-controlled "return_to" param
// OAuth-style callback redirect without validation
func (as *AuthServer) OAuthCallbackHandler(w http.ResponseWriter, r *http.Request) {
	code := r.URL.Query().Get("code")
	returnTo := r.URL.Query().Get("return_to")
	// VULN: return_to is user-controlled and used in redirect without validation
	// Attacker crafts link: /oauth/callback?code=xxx&return_to=https://evil.com
	if code != "" {
		if returnTo != "" {
			http.Redirect(w, r, returnTo, http.StatusFound)
		} else {
			http.Redirect(w, r, "/settings", http.StatusFound)
		}
	} else {
		http.Error(w, "missing code", 400)
	}
}

// VULN-3: Open redirect via http.Redirect with user-controlled "redirect" param
// Language switching endpoint used as open redirect vector
func (as *AuthServer) SetLanguageHandler(w http.ResponseWriter, r *http.Request) {
	lang := r.URL.Query().Get("lang")
	redirect := r.URL.Query().Get("redirect")
	// VULN: redirect param is user-controlled and used directly
	// e.g., /set-lang?lang=en&redirect=http://evil.com
	http.SetCookie(w, &http.Cookie{
		Name:  "lang",
		Value: lang,
		Path:  "/",
	})
	if redirect != "" {
		http.Redirect(w, r, redirect, http.StatusFound)
	} else {
		http.Redirect(w, r, "/", http.StatusFound)
	}
}

// SAFE: Open redirect protected by relative-path-only validation
// Only allows redirects to paths starting with "/" (no scheme/host)
func (as *AuthServer) SafePostLoginRedirectHandler(w http.ResponseWriter, r *http.Request) {
	redirectURL := r.FormValue("redirect")
	// SAFE: only allow relative redirects (path must start with / and not contain ://)
	if redirectURL != "" {
		parsed, err := url.Parse(redirectURL)
		if err != nil || parsed.IsAbs() || strings.HasPrefix(redirectURL, "//") {
			redirectURL = "/dashboard"
		}
		// Additional check: ensure it's a relative path
		if !strings.HasPrefix(redirectURL, "/") {
			redirectURL = "/dashboard"
		}
	} else {
		redirectURL = "/dashboard"
	}
	http.Redirect(w, r, redirectURL, http.StatusFound)
}

// SAFE: Open redirect protected by URL whitelist
// Only pre-approved redirect destinations are allowed
func (as *AuthServer) SafePaymentReturnHandler(w http.ResponseWriter, r *http.Request) {
	returnURL := r.URL.Query().Get("return")
	// SAFE: whitelist of allowed redirect destinations
	allowedReturns := map[string]bool{
		"/order/complete":  true,
		"/order/cancelled": true,
		"/checkout":         true,
		"/cart":             true,
	}
	if !allowedReturns[returnURL] {
		returnURL = "/order/complete"
	}
	http.Redirect(w, r, returnURL, http.StatusFound)
}

func main() {
	as := &AuthServer{}
	mux := http.NewServeMux()
	mux.HandleFunc("/auth/login", as.LoginRedirectHandler)
	mux.HandleFunc("/auth/oauth/callback", as.OAuthCallbackHandler)
	mux.HandleFunc("/set-lang", as.SetLanguageHandler)
	mux.HandleFunc("/auth/post-login", as.SafePostLoginRedirectHandler)
	mux.HandleFunc("/payment/return", as.SafePaymentReturnHandler)
	fmt.Println("Auth server running on :8086")
	http.ListenAndServe(":8086", mux)
}

File: tests/nodejs/realworld_open_redirect.js
/**
 * Real-world Open Redirect test cases for Kunlun-M scanner
 * Tests: res.redirect() with user-controlled URL from query/body/params
 */

const express = require('express');
const router = express.Router();

router.use(express.json());
router.use(express.urlencoded({ extended: true }));

/**
 * VULN: Open redirect via res.redirect with req.query URL
 * Attacker can redirect users to phishing sites
 * Example: /redirect?url=https://evil.com/phishing
 */
router.get('/redirect', (req, res) => {
    const targetUrl = req.query.url;
    // VULN: Unvalidated user-controlled URL passed directly to redirect
    res.redirect(targetUrl);
});

/**
 * VULN: Open redirect via res.redirect with req.body next parameter
 * Common in login flows where attacker provides a malicious "next" URL
 * Example POST: {"next": "https://attacker.com/fake-login"}
 */
router.post('/login', (req, res) => {
    const { username, password, next } = req.body;
    // Simulated auth check
    if (username === 'admin' && password === 'admin123') {
        // VULN: Redirect to user-supplied URL without validation
        res.redirect(next || '/dashboard');
    } else {
        res.status(401).send('Invalid credentials');
    }
});

/**
 * VULN: Open redirect via res.redirect with req.params with // bypass
 * Some frameworks strip leading // but not all
 * Example: /goto/https://evil.com (or //evil.com via protocol-relative)
 */
router.get('/goto/:target', (req, res) => {
    const target = req.params.target;
    // VULN: User-controlled target used in redirect without validation
    res.redirect(target);
});

/**
 * SAFE: Relative redirect only - use a whitelist of allowed paths
 * Prevents redirection to external domains
 */
router.get('/safe-redirect', (req, res) => {
    const target = req.query.url;
    // SAFE: Only allow relative redirects (starting with /)
    // Block any URL with a protocol or hostname
    if (!target || target.startsWith('http://') || target.startsWith('https://') ||
        target.startsWith('//') || target.startsWith('\\\\')) {
        return res.redirect('/home');
    }
    // Additionally verify it starts with / and is a known safe path
    const safePaths = ['/home', '/dashboard', '/profile', '/settings'];
    if (safePaths.some(p => target.startsWith(p))) {
        res.redirect(target);
    } else {
        res.redirect('/home');
    }
});

/**
 * SAFE: URL whitelist validation before redirect
 * Only specific pre-approved URLs are allowed as redirect destinations
 */
router.post('/safe-login', (req, res) => {
    const { username, password, next } = req.body;
    if (username === 'admin' && password === 'admin123') {
        // SAFE: Whitelist of allowed redirect destinations
        const allowedRedirects = ['/dashboard', '/home', '/profile', '/settings'];
        const redirectTarget = allowedRedirects.includes(next) ? next : '/dashboard';
        res.redirect(redirectTarget);
    } else {
        res.status(401).send('Invalid credentials');
    }
});

module.exports = router;
